src/gls/server.c3

```diff
@@ -0,0 +1,763 @@
+        let resp : StringView = sv("{\"jsonrpc\":\"2.0\",\"method\":\"textDocument/publishDiagnostics\",\"params\":{\"uri\": \"%\", \"version\": %, \"diagnostics\": []}}");
+        mut ss : StringStream;
+        ss.buffer_pointer = 0;
+        ss:writef(resp, uri:data().value:data(), version:data().value:data());
+        print(ss:as_sv());
+        out:writef(sv("Content-Length: %\r\n"), ss:length());
+        out:write(sv("\r\n"));
+        out:write(ss:as_sv());
+        out:flush();
+        json:data():deallocate();
+        return;
+    }
+
+    // TODO need an actual API here.
+    if success:data().type != /* false */ 2 {
+        json:data():deallocate();
+        return;
+    }
+
+    let message : Optional<JSON_Node> = json:data().map:at(sv("message"));
+    if !message:has_value() {
+        print(sv("'message' not found"));
+        json:data():deallocate();
+        return;
+    }
+
+    let location : Optional<JSON_Node> = json:data().map:at(sv("location"));
+    if !location:has_value() {
+        print(sv("'location' not found"));
+        json:data():deallocate();
+        return;
+    }
+
+    let line : Optional<JSON_Node> = location:data().map:at(sv("line"));
+    if !line:has_value() {
+        print(sv("'line' not found"));
+        json:data():deallocate();
+        return;
+    }
+
+    let column : Optional<JSON_Node> = location:data().map:at(sv("column"));
+    if !column:has_value() {
+        print(sv("'column' not found"));
+        json:data():deallocate();
+        return;
+    }
+
+    mut cursor : Cursor = { line:data().value:data():get_iter(), &mut state.allocator };
+    let line_number : Optional<int> = cursor:parse_natural_number();
+
+    if !line_number:has_value() {
+        print(sv("Couldn't parse line number"));
+        json:data():deallocate();
+        return;
+    }
+
+    cursor.iter = column:data().value:data():get_iter();
+    let column_number : Optional<int> = cursor:parse_natural_number();
+
+    if !column_number:has_value() {
+        print(sv("Couldn't parse column number"));
+        json:data():deallocate();
+        return;
+    }
+
+    printf(sv("message: %, line: %"), message:data().value:data(), line:data().value:data());
+    let resp : StringView = sv("{\"jsonrpc\":\"2.0\",\"method\":\"textDocument/publishDiagnostics\",\"params\":{\"uri\": \"%\", \"version\": %, \"diagnostics\": [{ \"range\": { \"start\": { \"line\": %, \"character\": % }, \"end\" : { \"line\": %, \"character\": % } }, \"message\": \"%\" }]}}");
+    mut ss : StringStream;
+    ss.buffer_pointer = 0;
+    ss:writef(resp, uri:data().value:data(), version:data().value:data(), line_number:data() - 1, column_number:data() - 1, line_number:data() - 1, column_number:data() - 1, message:data().value:data());
+    print(ss:as_sv());
+    out:writef(sv("Content-Length: %\r\n"), ss:length());
+    out:write(sv("\r\n"));
+    out:write(ss:as_sv());
+    out:flush();
+
+    json:data():deallocate();
+}
```

Comment:
The `diagnostic` and `file_changed` sections seem quite similar. Have you tried reusing the logic (especially the parser output parsing).

Reply:
I've cleaned this up a bit